### PR TITLE
feat(typescript): Upgrade to TS v4 (back compatible with TS v3.8)

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :loudspeaker: Type of change

- [x] Enhancement

## :scroll: Description

The library compiles using TS v4 and from v4 declaration files, v3.8 files are created using downlevel-dts. 
Library ships with both the new and old types.
CI checks if the downleveled types are working correctly with 3.8 tsc.

## :bulb: Motivation and Context

closes: https://github.com/getsentry/sentry-react-native/issues/2983

## :green_heart: How did you test it?
ci, test project

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: src/js/integrations/sdkinfo.ts [hunk 1]
```diff
@@ -64,3 +64,12 @@ export class SdkInfo implements Integration {
...
 }
+
+//this is 4.0 feature and cant be downleveled
+type Strings = [string, string];
+type Numbers = [number, number];
+export type StrStrNumNumBool = [...Strings, ...Numbers, boolean];
+
+//this will be downleveled
+type World = "world";
+export type Greeting = `hello ${World}`;
```
Comment (krystofwoldrich) [inline src/js/integrations/sdkinfo.ts:75]: This is an example of code that will be downleveled and code that can not be downleveled. When CI is done, I will link its output.

File: test/typescript/ts3.8/index.ts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+import * as _Sentry from '@sentry/react-native';
+// we need to import the SDK to ensure tsc check the types
```
Comment (marandaneto) [inline test/typescript/ts3.8/index.ts:2]:
```suggestion
// we need to import the SDK to ensure tsc check the types
import * as _Sentry from '@sentry/react-native';
```

**Review by marandaneto — APPROVED** [review]
E2E test is unhappy, otherwise LGTM.

**Comment by krystofwoldrich** [comment]
@marandaneto Thanks check it, seems just like flakiness, `yarn install` failed. I'll re-run it.

**Pull request merged**